exercise/cache/points.py

```diff
@@ -274,58 +348,69 @@ def get_max_deviations(
                 )
 
             # Augment exercise reveal rules.
-            if not self.is_staff:
-                for exercise in (
-                    BaseExercise.objects
-                    .filter(course_module__course_instance=instance)
-                    .prefetch_related('submission_feedback_reveal_rule')
-                    .only('id', 'submission_feedback_reveal_rule')
-                ):
-                    try:
-                        tree = self._by_idx(modules, exercise_index[exercise.id])
-                    except KeyError:
-                        self.dirty = True
-                        continue
-                    entry = tree[-1]
-                    rule = exercise.active_submission_feedback_reveal_rule
-                    state = ExerciseRevealState(entry)
-                    is_revealed = rule.is_revealed(state)
-                    reveal_time = rule.get_reveal_time(state)
-
+            for exercise in (
+                BaseExercise.objects
+                .filter(course_module__course_instance=instance)
+                .prefetch_related('submission_feedback_reveal_rule')
+                .only('id', 'submission_feedback_reveal_rule')
+            ):
+                try:
+                    tree = self._by_idx(modules, exercise_index[exercise.id])
+                except KeyError:
+                    self.dirty = True
+                    continue
+                entry = tree[-1]
+                rule = exercise.active_submission_feedback_reveal_rule
+
+                # Evaluate the reveal rule using the real (staff) values.
+                entry_copy = dict(entry)
+                del entry_copy['submissions'] # Not needed by the reveal rule and we don't want to unprefix them.
+                self._unprefix(entry_copy, True)
+                state = ExerciseRevealState(entry_copy)
+                is_revealed = rule.is_revealed(state)
+                reveal_time = rule.get_reveal_time(state)
+
+                entry.update({
+                    '_student_feedback_revealed': is_revealed,
+                    '_student_feedback_reveal_time': reveal_time,
+                })
+                if is_revealed:
                     entry.update({
-                        'points': entry['points'] if is_revealed else 0,
-                        'formatted_points': format_points(entry['points'], is_revealed, False),
-                        'passed': entry['passed'] if is_revealed else False,
-                        'feedback_revealed': is_revealed,
-                        'feedback_reveal_time': reveal_time,
+                        '_student_best_submission': entry['_staff_best_submission'],
```

Comment:
So student and staff keys are mixed here because now the student value must be the fully revealed value, which has been previously stored under the staff key, right?

Reply:
Yes, and this should also answer some of your earlier questions. Initially, the code assumes that feedback is not revealed, and stores the exercise results (points, passed, best_submission, etc) under the staff keys **only**. And here we check if feedback has been revealed, and if it is, we copy the exercise results from the staff keys to the student keys.

The reason to do it in this order is that the we must have constructed some kind of a cache before we can evaluate reveal rules (see the comment related to lines 366-376).